packages/ui-keyring/src/index.ts

```diff
@@ -4,83 +4,429 @@
-  isAvailable: (address: string | Uint8Array): boolean =>
-    isAvailable(state, address),
-  isPassValid: (password: string): boolean =>
-    isPassValid(state, password),
-  getAccounts: (): Array<KeyringAddress> =>
-    getAccounts(state),
-  getAddress: (address: string | Uint8Array): KeyringAddress =>
-    getAddress(state, address),
-  getAddresses: (): Array<KeyringAddress> =>
-    getAddresses(state),
-  getPair: (address: string | Uint8Array): KeyringPair =>
-    state.keyring.getPair(address),
-  getPairs: (): Array<KeyringPair> =>
-    state.keyring.getPairs().filter((pair) =>
-      development.isDevelopment() || pair.getMeta().isTesting !== true
-    ),
-  loadAll: (): void =>
-    loadAll(state),
-  restoreAccount: (json: KeyringPair$Json, password: string): KeyringPair =>
-    restoreAccount(state, json, password),
-  saveAccount: (pair: KeyringPair, password?: string): void =>
-    saveAccount(state, pair, password),
-  saveAccountMeta: (pair: KeyringPair, meta: KeyringPair$Meta): void =>
-    saveAccountMeta(state, pair, meta),
-  saveAddress: (address: string, meta: KeyringPair$Meta): void =>
-    saveAddress(state, address, meta),
-  saveRecent: (address: string): SingleAddress =>
-    saveRecent(state, address),
-  setDevMode: (isDevelopment: boolean): void =>
-    development.set(isDevelopment)
-} as KeyringInstance);
+class Keyring implements KeyringInstance {
+  private state: State;
+  private id: number = 0;
+
+  static counter: number;
+
+  constructor () {
+    // state must be always defined in the constructor even if trying to create second instance to overcome TSLint error
+    this.state = {
+      accounts: observableAccounts,
+      addresses: observableAddresses,
+      keyring: testKeyring()
+    };
+
+    if (!singletonInstance) {
+      singletonInstance = this;
+    } else {
+      return;
+    }
+
+    // NOTE Everything is loaded in API after chain is received
+    // this.loadAll();
+
+    this.id = ++Keyring.counter;
+    assert(this.id <= 1, 'KeyringInstance should be singleton');
+
+    return singletonInstance;
+  }
+
+  private addAccountPairs (): void {
+    const { accounts, keyring } = this.state;
+
+    keyring
+      .getPairs()
+      .forEach((pair) => {
+        const address = pair.address();
+
+        accounts.add(address, {
+          address,
+          meta: pair.getMeta()
+        });
+      });
+  }
+
+  private addAddresses (options: KeyringOptions): void {
+    const { addresses } = this.state;
+    const addressesAvailable = addresses.subject.getValue();
+
+    Object
+      .keys(addressesAvailable)
+      .map((address) =>
+        addressesAvailable[address]
+      )
+      .forEach(({ json: { meta: { isRecent = false } }, option }: SingleAddress) => {
+        if (isRecent) {
+          options.recent.push(option);
+        } else {
+          options.address.push(option);
+        }
+      });
+  }
+
+  private emptyOptions (): KeyringOptions {
```

Comment:
I would suggest (as per the comments above), to just keep this to the add, remove, etc. logic - rather keep the observable stuff seperate?